solr/contrib/ltr/src/java/org/apache/solr/ltr/response/transform/LTRFeatureLoggerTransformerFactory.java

```diff
@@ -208,55 +216,116 @@ public void setContext(ResultContext context) {
+    }
 
-      try {
-        modelWeight = scoringQuery.createWeight(searcher, ScoreMode.COMPLETE, 1f);
-      } catch (final IOException e) {
-        throw new SolrException(SolrException.ErrorCode.BAD_REQUEST, e.getMessage(), e);
+    /**
+     * When preparing the reranking queries for logging features various scenarios apply:
+     * 
+     * No Reranking 
+     * There is the need of a logger model from the default feature store/ the explicit feature store passed
+     * to extract the feature vector
+     * 
+     * Re Ranking
+     * 1) If no explicit feature store is passed, the models for each reranking query can be safely re-used
+     * the feature vector can be fetched from the feature vector cache.
+     * 2) If an explicit feature store is passed, and no reranking query uses a model from that featureStore,
+     * There is the need of a logger model to extract the feature vector
+     * 3) If an explicit feature store is passed, and there is a reranking query that uses a model from that featureStore,
+     * It can be re-used
+     * 
+     * @param rerankingQueriesFromContext reranking queries
+     * @param transformerFeatureStore explicit feature store for the transformer
+     * @param transformerExternalFeatureInfo explicit efi for the transformer
+     */
+    private void setupRerankingQueriesForLogging(LTRScoringQuery[] rerankingQueriesFromContext, String transformerFeatureStore, Map<String, String[]> transformerExternalFeatureInfo) {
+      if (docsWereNotReranked) { //no reranking query
+        LTRScoringQuery loggingQuery = new LTRScoringQuery(loggingModel,
+            transformerExternalFeatureInfo,
+            true,
+            threadManager);
+        rerankingQueries = new LTRScoringQuery[]{loggingQuery};
+      } else {
+        rerankingQueries = new LTRScoringQuery[rerankingQueriesFromContext.length];
+        System.arraycopy(rerankingQueriesFromContext, 0, rerankingQueries, 0, rerankingQueriesFromContext.length);
+
+        if (transformerFeatureStore != null) {// explicit feature store for the transformer
+          LTRScoringModel matchingRerankingModel = null;
+          for (LTRScoringQuery rerankingQuery : rerankingQueries) {
+            if (isModelMatchingFeatureStore(transformerFeatureStore, rerankingQuery.getScoringModel())) {
+              matchingRerankingModel = rerankingQuery.getScoringModel();
+            }
+          }
+          if (matchingRerankingModel != null) {//one of the LTR query model can be re-used
+            for (LTRScoringQuery rerankingQuery : rerankingQueries) {
+              rerankingQuery.setLtrScoringModel(matchingRerankingModel);
+            }
+          } else {
+            for (LTRScoringQuery rerankingQuery : rerankingQueries) {
+              rerankingQuery.setLtrScoringModel(loggingModel);
+              rerankingQuery.setExtractAllFeatures(true);
+              rerankingQuery.setEfi(transformerExternalFeatureInfo);
+            }
```

Comment:
LTRFeatureLoggerTransformerFactory.6 - the reranking queries from the context were copied and now they are being modified. The "why copy-and-modify instead construct-a-new-one?" question may arise for code readers.

Reply:
LTRScoringQuery queries may be LTRInterleavingScoringQuery, containing the picked document Ids we don't want to lose

Reply:
but this is solved with the queries from context :)